packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -151,16 +150,22 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
     private readonly contextsDeferred = new Map<string, Deferred<IChannelContext>>();
     private readonly pendingAttach = new Map<string, IAttachMessage>();
 
-    private bindState: BindState;
-    // For new data stores, this is used to break the recursion while attaching the graph. The graph must be attached
-    // before the data store can move to Attached state (see _attachState) and become live.
-    // For existing data stores, the graph is always attached.
-    private graphAttachState: AttachState = AttachState.Detached;
+    /**
+     * Tracks whether this data store is locally attached to the container's object graph. Being locally attached makes
+     * it visible in the container via the root of the graph.
+     * The data store must be locally attached before it can become live and be visible globally. In other words, if
+     * the data store is globally attached, it has to be locally attached as well.
+     */
+    private localAttachState: AttachState;
+    /**
+     * Tracks whether this data store is attached globally and is visible to all other clients. It must be locally
+     * attached before it can be globally attached.
+     */
+    private _globalAttachState: AttachState;
```

Comment:
One thing that I do not like is possibility of states that do not make sense. For example _globalAttachState == Attached & localAttachState !== Attached. I'd rather have one enum that clearly spells actual states.

Reply:
Yeah, having one enum with these 3 states makes sense. But that will make this an even bigger change because the "attachState" is exposed in `IDataStoreRuntime` interface. So, we can try to simplify that part in a follow up.

Reply:
For now, I removed the "attaching" state from data store runtime. I don't see why data store context needs attaching state either but I am not touching the context in this change.